ipa-core/src/report.rs

```diff
@@ -413,6 +434,186 @@ where
+
+    const TS_OFFSET: usize = 0;
+
+    const BK_OFFSET: usize = Self::TS_OFFSET + <Replicated<TS> as Serializable>::Size::USIZE;
+    const TV_OFFSET: usize = Self::BK_OFFSET + <Replicated<BK> as Serializable>::Size::USIZE;
+
+    pub fn encap_key_mk(&self) -> &[u8] {
+        &self.data[Self::ENCAP_KEY_MK_OFFSET..Self::CIPHERTEXT_MK_OFFSET]
+    }
+
+    pub fn mk_ciphertext(&self) -> &[u8] {
+        &self.data[Self::CIPHERTEXT_MK_OFFSET..Self::ENCAP_KEY_BTT_OFFSET]
+    }
+
+    pub fn encap_key_btt(&self) -> &[u8] {
+        &self.data[Self::ENCAP_KEY_BTT_OFFSET..Self::CIPHERTEXT_BTT_OFFSET]
+    }
+
+    pub fn btt_ciphertext(&self) -> &[u8] {
+        &self.data[Self::CIPHERTEXT_BTT_OFFSET..Self::EVENT_TYPE_OFFSET]
+    }
+
+    /// ## Panics
+    /// Only if a `Report` constructor failed to validate the contents properly, which would be a bug.
+    pub fn event_type(&self) -> EventType {
+        EventType::try_from(self.data[Self::EVENT_TYPE_OFFSET]).unwrap() // validated on construction
+    }
+
+    pub fn key_id(&self) -> KeyIdentifier {
+        self.data[Self::EVENT_TYPE_OFFSET + 1]
+    }
+
+    /// ## Panics
+    /// Never.
+    pub fn epoch(&self) -> Epoch {
+        u16::from_le_bytes(
+            self.data[Self::EVENT_TYPE_OFFSET + 2..Self::SITE_DOMAIN_OFFSET]
+                .try_into()
+                .unwrap(), // infallible slice-to-array conversion
+        )
+    }
+
+    /// ## Panics
+    /// Only if a `Report` constructor failed to validate the contents properly, which would be a bug.
+    pub fn site_domain(&self) -> &str {
+        std::str::from_utf8(&self.data[Self::SITE_DOMAIN_OFFSET..]).unwrap() // validated on construction
+    }
+
+    /// ## Errors
+    /// If the report contents are invalid.
+    pub fn from_bytes(bytes: B) -> Result<Self, InvalidReportError> {
+        EventType::try_from(bytes[Self::EVENT_TYPE_OFFSET])?;
+        let site_domain = &bytes[Self::SITE_DOMAIN_OFFSET..];
+        if !site_domain.is_ascii() {
+            return Err(NonAsciiStringError::from(site_domain).into());
+        }
+        Ok(Self {
+            data: bytes,
+            phantom_data: PhantomData,
+        })
+    }
+
+    /// ## Errors
+    /// If the match key shares in the report cannot be decrypted (e.g. due to a
+    /// failure of the authenticated encryption).
+    /// ## Panics
+    /// Should not panic. Only panics if a `Report` constructor failed to validate the
+    /// contents properly, which would be a bug.
+    pub fn decrypt(
+        &self,
+        key_registry: &KeyRegistry<KeyPair>,
+    ) -> Result<OprfReport<BK, TV, TS>, InvalidReportError> {
+        let info = Info::new(
+            self.key_id(),
+            self.epoch(),
+            self.event_type(),
+            HELPER_ORIGIN,
+            self.site_domain(),
+        )
+        .unwrap(); // validated on construction
+
+        let mut ct_mk = self.mk_ciphertext().to_vec();
```

Comment:
can we keep it on the stack without requiring an extra allocation per report?

Reply:
Yes we don't need that, again was following Andy here. I assume he made the extra copy since during the next step, open_in_place, we would destroy the original mk_ciphertext. I don't think we need it for anything, though.

Reply:
we would need to change the function from &self to &mut self. I suggest to leave it for now.

Reply:
Can you point me to the heap allocation in the old code? I am struggling to find it - it does seem that it works w/o a roundtrip to jemalloc.

The reason I think it matters is because this will be called 50 million (or whatever shard size is) times - we got a decent chance for memory allocator to do its job properly, but this is a game I think we don't have to play as it should be possible to avoid in this code

Reply:
@danielmasny pointed that w/o const generic arithmetic we can't really create a generic array of size MK+PK+TAG w/o extra trait bounds `Sum<Add>`.... Before we let this code to go, I wonder if @andyleiserson has any suggestions here? One optimization I can think of is to allocate outside and pass the buffer inside the `decrypt` method, but jemalloc is probably doing it already for us

Reply:
I fixed the ciphertext length. The trait bounds are not too bad anymore since you were right, I don't need to round up to a multiple of 16 bytes.